packages/clay-data-provider/src/useResource.ts

```diff
@@ -0,0 +1,284 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import LRU from 'lru-cache';
+import {
+	FetchPolicy,
+	IDataProvider,
+	NetworkStatus,
+	SYMBOL_DATA_PROVIDER,
+	SYMBOL_ORIGIN,
+	TSymbolData,
+} from './types';
+import {timeout} from './util';
+import {useDebounce} from './useDebounce';
+import {useEffect, useRef, useState} from 'react';
+
+interface IResource extends IDataProvider {
+	onNetworkStatusChange?: (status: NetworkStatus) => void;
+}
+
+const evaluteVariables = (variables: any): string | null => {
+	if (!variables) {
+		return null;
+	}
+
+	const keys = Object.keys(variables);
+
+	return keys
+		.map((key, index) => {
+			const param = `${key}=${variables[key]}`;
+			return index === keys.length - 1 ? `?${param}` : param;
+		})
+		.join('&');
+};
+
+const useResource = ({
+	fetchDelay = 300,
+	fetchOptions,
+	fetchPolicy = FetchPolicy.NoCache,
+	fetchTimeout = 6000,
+	link,
+	onNetworkStatusChange = () => {},
+	pollInterval = 0,
+	storage = {
+		/**
+		 * This will ensure that we know that the storage
+		 * reference is not external, otherwise the cache
+		 * will persist by the application.
+		 */
+		[SYMBOL_ORIGIN]: true,
+	},
+	fetchRetry = {
+		delay: {
+			initial: 300,
+			jitter: true,
+		},
+		attempts: 5,
+	},
+	storageMaxSize = 20,
+	variables = null,
+}: IResource) => {
+	const [resource, setResource] = useState<any>(null);
+
+	const getStorageCache = () => {
+		if (storage[SYMBOL_DATA_PROVIDER]) {
+			return storage[SYMBOL_DATA_PROVIDER] as TSymbolData;
+		}
+
+		return (storage[SYMBOL_DATA_PROVIDER] = new LRU(storageMaxSize));
+	};
+
+	let networkStatus = useRef<NetworkStatus>(NetworkStatus.Unused).current;
+
+	let pollingIntervalIdRef = useRef<null | NodeJS.Timeout>(null).current;
```

Comment:
Since we access `.current`, we should probably remove the `ref` from the variable declaration. I think it makes most sense for all ` **Ref` variables to be the ref object itself.

Reply:
Hm yeah, makes sense.